diff --git a/kvm/x86/cpuid.c b/kvm/x86/cpuid.c
index 34446db..bfc5a7d 100644
--- a/kvm/x86/cpuid.c
+++ b/kvm/x86/cpuid.c
@@ -133,7 +133,7 @@ int kvm_update_cpuid(struct kvm_vcpu *vcpu)
 	if (best && (best->eax & (F(XSAVES) | F(XSAVEC))))
 		best->ebx = xstate_required_size(vcpu->arch.xcr0, true);
 
-	vcpu->arch.eager_fpu = use_eager_fpu() || guest_cpuid_has_mpx(vcpu);
+	vcpu->arch.eager_fpu = guest_cpuid_has_mpx(vcpu);
 	if (vcpu->arch.eager_fpu)
 		kvm_x86_ops->fpu_activate(vcpu);
 
